### PR TITLE
Port of framework_agreement module

### PR DESCRIPTION
**Pull request by nbessi**

### CODE REVIEW

**Comment by lepistone** [comment]
Upstream had two lint commits. I resolved the conflicts automatically with `git rebase -m -X theirs 8.0`

**Comment by lepistone** [comment]
I fixed a few things myself. Thanks @nbessi!
:+1:

**Comment by eLBati** [comment]
:+1:

**Comment by jgrandguillaume** [comment]
Hi,

I've tested the module, some remarks:
- The name of the module in the **openerp**.py should be "Framework agreement" not "SImple Framework agrement"
- the field partner in the framework agreement should have a domain on partner "supplier=True"
- When I create a PO for camptocamp and try to add a PO line, I get the following error [1]

I wasn't able to test the features further because of this...

Thanks for the work,

Joël

[1] : 

```
Traceback (most recent call last):
  File ".../parts/server/openerp/http.py", line 500, in _handle_exception
    return super(JsonRequest, self)._handle_exception(exception)
  File ".../parts/server/openerp/http.py", line 517, in dispatch
    result = self._call_function(**self.params)
  File ".../parts/server/openerp/http.py", line 283, in _call_function
    return checked_call(self.db, *args, **kwargs)
  File ".../parts/server/openerp/service/model.py", line 113, in wrapper
    return f(dbname, *args, **kwargs)
  File ".../parts/server/openerp/http.py", line 280, in checked_call
    return self.endpoint(*a, **kw)
  File ".../parts/server/openerp/http.py", line 733, in __call__
    return self.method(*args, **kw)
  File ".../parts/server/openerp/http.py", line 376, in response_wrap
    response = f(*args, **kw)
  File ".../parts/server/addons/web/controllers/main.py", line 944, in call_kw
    return self._call_kw(model, method, args, kwargs)
  File ".../parts/server/addons/web/controllers/main.py", line 936, in _call_kw
    return getattr(request.registry.get(model), method)(request.cr, request.uid, *args, **kwargs)
  File ".../parts/server/openerp/api.py", line 237, in wrapper
    return old_api(self, *args, **kwargs)
  File ".../parts/server/openerp/api.py", line 360, in old_api
    result = method(recs, *args, **kwargs)
  File ".../parts/server/openerp/models.py", line 5703, in onchange
    record._onchange_eval(name, field_onchange[name], result)
  File ".../parts/server/openerp/models.py", line 5637, in _onchange_eval
    method_res = getattr(self._model, method)(*args)
  File ".../parts/server/openerp/api.py", line 237, in wrapper
    return old_api(self, *args, **kwargs)
TypeError: onchange_product_id() takes at most 16 arguments (17 given)
```

**Comment by lepistone** [comment]
I found out that while travis is green, tests fail on my machine. I am investigating that and @jgrandguillaume's remarks.

**Comment by lepistone** [comment]
note: because of https://github.com/odoo/odoo/issues/3056 (and another bug as well, which I didn't report because it was fixed today), tests have to be run with odoo at commit https://github.com/odoo/odoo/commit/c1d48362d3a38606ffb99d569fdc6147dcd0c775

**Comment by lepistone** [comment]
Hi @jgrandguillaume, did you install other modules alongside framework_agreement? I don't get to reproduce it.

On the other hand, the solution seems to be to use `web_context_tunnel` because framework_agreement does add a parameter to an onchange, and that can interfere with other modules overriding the same onchange.

**Comment by jgrandguillaume** [comment]
Hi,

I trash my DB and restart fresh with the following modules and still have the same trouble : 

Technical Name
crm
mail
account_voucher
sale
stock
stock_account
im_chat
purchase
account
analytic
auth_crypt
auth_signup
base
base_action_rule
base_import
base_setup
board
bus
calendar
claim_from_delivery
crm_claim
decimal_precision
edi
email_template
fetchmail
framework_agreement
im_odoo_support
payment
payment_transfer
portal
portal_claim
portal_sale
portal_stock
procurement
product
purchase_requisition
purchase_requisition_bid_selection
purchase_rfq_bid_workflow
report
resource
sale_crm
sale_journal
sale_sourced_by_line
sales_team
sale_stock
share
stock_dropshipping
stock_picking_wave
web
web_calendar
web_context_tunnel
web_diagram
web_gantt
web_graph
web_kanban
web_kanban_gauge
web_kanban_sparkline
web_tests
web_view_editor

**Comment by lepistone** [comment]
Hi @jgrandguillaume, I fixed the problems you reported and adapted the module to web_context_tunnel, so that it does not change the signature of on_change methods anymore.

Tests are green if run locally at https://github.com/odoo/odoo/commit/bc59cfce618a974600dec384b6c41ee0b186feed

**Comment by lepistone** [comment]
I changed the minimum I could to work around odoo/odoo#3056 so we're green! As @jgrandguillaume remarks should be fixed with `web_context_tunnel` I think we can merge.

**Comment by jgrandguillaume** [comment]
Hi,

Ok, the previous trouble are fixed. I now face another trouble:

1) Create an agreement for Camptocamp, available from today -1 to next year, for product A, qty 1000, price 100.-
2) Create a PO for camptocamp with agreement above
3) Add a line with product A for 10 pce
4) The unit price in the line isn't computed according to the framework agreement
5) Confirm the PO, the framework agreement decrease the quantity of the LTA correctly

=> Can you check why the price isn't computed and why the tests seems to be green !? We want a test that prevent this trouble with the fix please.

Thanks,

Joël

**Comment by lepistone** [comment]
ok @jgrandguillaume, I'm giving a look

**Comment by bwrsandman** [comment]
Code review :+1:

**Comment by lepistone** [comment]
Hi @jgrandguillaume: I found and fixed a bug that probably was not the one you're referring to: the onchange that prevents changes to the unit price on PO lines was not being called from the view. (to make an integration test that checks this warning, I would need to assertRaises in a yaml !record that runs onchange methods automatically, which I can't do easily AFAIK)

On the other hand, I do not reproduce your situation: In my RFQ I see the price from the agreement.

As it is implemented with pricelists, maybe there is an interaction with another module that touches pricelists?

Thanks!

**Comment by jgrandguillaume** [comment]
Hi @lepistone 

Everything is fine now, thank you !

We can merge now :+1: 

Regards,

File: framework_agreement/__openerp__.py [hunk 1]
```diff
@@ -2,7 +2,7 @@
 ##############################################################################
 #
 #    Author: Nicolas Bessi
-#    Copyright 2013 Camptocamp SA
+#    Copyright 2014 Camptocamp SA
```
Comment (yvaucher) [inline framework_agreement/__openerp__.py]: Shouldn't it be 2013-2014 ?

**Comment by yvaucher** [comment]
:+1:

**Pull request merged**